ledger/acctonline.go

```diff
@@ -0,0 +1,863 @@
+		if !errors.As(err, &roundOffsetError) {
+			return
+		}
+		// the round number cannot be found in deltas, it is in history
+		inHistory = true
+		err = nil
+	}
+	paramsOffset, err = ao.roundParamsOffset(rnd)
+	if err != nil {
+		return
+	}
+
+	rewardsProto = config.Consensus[ao.onlineRoundParamsData[paramsOffset].CurrentProtocol]
+	rewardsLevel = ao.onlineRoundParamsData[paramsOffset].RewardsLevel
+
+	// check if we've had this address modified in the past rounds. ( i.e. if it's in the deltas )
+	if !inHistory {
+		macct, indeltas := ao.accounts[addr]
+		if indeltas {
+			// Check if this is the most recent round, in which case, we can
+			// use a cache of the most recent account state.
+			if offset == uint64(len(ao.deltas)) {
+				return macct.data.OnlineAccountData(rewardsProto, rewardsLevel), nil
+			}
+			// the account appears in the deltas, but we don't know if it appears in the
+			// delta range of [0..offset], so we'll need to check :
+			// Traverse the deltas backwards to ensure that later updates take
+			// priority if present.
+			for offset > 0 {
+				offset--
+				d, ok := ao.deltas[offset].GetData(addr)
+				if ok {
+					return d.OnlineAccountData(rewardsProto, rewardsLevel), nil
+				}
+			}
+		}
+	}
+
+	if macct, has := ao.onlineAccountsCache.read(addr, rnd); has {
+		return macct.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
+	}
+
+	ao.accountsMu.RUnlock()
+	needUnlock = false
+
+	// No updates of this account in the in-memory deltas; use on-disk DB.
+	// As an optimization, we avoid creating
+	// a separate transaction here, and directly use a prepared SQL query
+	// against the database.
+	persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
+	if err != nil {
+		return ledgercore.OnlineAccountData{}, err
+	}
+
+	if persistedData.rowid != 0 { // if we read actual data return it
+		// lookupOnlineHistory does not fetch the account db round because of the following observation:
+		// 1. ao.onlineAccountsCache update happens with ao.accountsMu taken below and in postCommit
+		// 2. If we started reading the history (lookupOnlineHistory)
+		//   * before commitRound or while it is running => OK, read what is in DB and then add new entries in postCommit
+		//     - if commitRound deletes some history after, the cache has additional entries and updRound comparison gets a right value
+		//   * after commitRound but before postCommit => OK, read full history, ignore the update from postCommit in writeFront's updRound comparison
+		//   * after postCommit => OK, postCommit does not add new entry with writeFrontIfExist, but here all the full history is loaded
+		persistedDataHistory, err := ao.accountsq.lookupOnlineHistory(addr)
+		if err != nil {
+			return ledgercore.OnlineAccountData{}, err
+		}
+		err = func() error {
+			ao.accountsMu.Lock()
+			defer ao.accountsMu.Unlock()
+			ao.onlineAccountsCache.clear(addr)
+			if !ao.onlineAccountsCache.full() {
+				for _, data := range persistedDataHistory {
```

Comment:
At this point `persistedDataHistory` can be too old. I think we need to compare the round of `persistedDataHistory` and `cachedDBRoundOnline` under this lock.

Reply:
I agree, if we finished reading, and something was committed but not written into the cache is `postCommit` is fine, but if `postCommit` is called (locked) before this helper function then yes, there is a problem, the last entry will be missed from the cache. I'll submit a fix.

Reply:
Fixed in https://github.com/algorand/go-algorand/pull/4038